pkg/workflow/controllers/statuscheck_reconciler.go

```diff
@@ -0,0 +1,270 @@
+)
+
+type StatusCheckReconciler struct {
+	kubeClient    client.Client
+	eventRecorder recorder.ChaosRecorder
+	logger        logr.Logger
+}
+
+func NewStatusCheckReconciler(kubeClient client.Client, eventRecorder recorder.ChaosRecorder, logger logr.Logger) *StatusCheckReconciler {
+	return &StatusCheckReconciler{kubeClient: kubeClient, eventRecorder: eventRecorder, logger: logger}
+}
+
+func (it *StatusCheckReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	startTime := time.Now()
+	defer func() {
+		it.logger.V(4).Info("finished syncing for status check node",
+			"node", request.NamespacedName,
+			"duration", time.Since(startTime),
+		)
+	}()
+
+	node := v1alpha1.WorkflowNode{}
+	err := it.kubeClient.Get(ctx, request.NamespacedName, &node)
+	if err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+	if node.Spec.Type != v1alpha1.TypeStatusCheck {
+		return reconcile.Result{}, nil
+	}
+
+	it.logger.V(4).Info("resolve status check node", "node", request)
+	if err := it.syncStatusCheck(ctx, request, node); err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "sync status check")
+	}
+
+	updateError := retry.RetryOnConflict(retry.DefaultRetry, it.updateNodeStatus(ctx, request))
+
+	return reconcile.Result{}, updateError
+}
+
+func (it *StatusCheckReconciler) syncStatusCheck(ctx context.Context, request reconcile.Request, node v1alpha1.WorkflowNode) error {
+	statusChecks, err := it.fetchChildrenStatusCheck(ctx, node)
+	if err != nil {
+		return errors.Wrap(err, "fetch children status check")
+	}
+
+	if WorkflowNodeFinished(node.Status) {
+		for _, item := range statusChecks {
+			// best efforts deletion
+			item := item
+			err := it.kubeClient.Delete(ctx, &item)
+			if client.IgnoreNotFound(err) != nil {
+				it.logger.Error(err, "failed to delete StatusCheck for workflow status check node",
+					"namespace", node.Namespace,
+					"nodeName", node.Name,
+					"statusCheckName", item.GetName(),
+				)
+				it.eventRecorder.Event(&node, recorder.StatusCheckDeletedFailed{Name: item.GetName()})
+			} else {
+				it.eventRecorder.Event(&node, recorder.StatusCheckDeleted{Name: item.GetName()})
+			}
+		}
+		return nil
+	}
+
+	if len(statusChecks) == 0 {
+		parentWorkflow, err := getParentWorkflow(ctx, it.kubeClient, node)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		spawnedStatusCheck, err := it.spawnStatusCheck(ctx, &node, parentWorkflow)
+		if err != nil {
+			it.eventRecorder.Event(&node, recorder.StatusCheckCreatedFailed{Name: spawnedStatusCheck.GetName()})
+			return errors.Wrap(err, "spawn status check")
+		}
+		it.eventRecorder.Event(&node, recorder.StatusCheckCreated{Name: spawnedStatusCheck.GetName()})
+	} else if len(statusChecks) > 1 {
+		var statusCheckToRemove []string
+		for _, item := range statusChecks[1:] {
+			statusCheckToRemove = append(statusCheckToRemove, item.GetName())
+		}
+		it.logger.Info("removing duplicated StatusCheck",
```

Comment:
why exist some duplicate status checks?

Reply:
It's the same reason with #3112 🤣.

Reply:
maybe we can update `duplicated` to `redundant` later